packages/automerge-repo/src/storage/StorageSubsystem.ts

```diff
@@ -1,97 +1,80 @@
     if (binary && binary.length > 0) {
-      if (!this.incrementalChanges[documentId]) {
-        this.incrementalChanges[documentId] = 0
+      if (!this.#changeCount[documentId]) {
+        this.#changeCount[documentId] = 0
       }
 
-      this.storageAdapter.save(
-        `${documentId}.incremental.${this.incrementalChanges[documentId]}`,
+      this.#storageAdapter.save(
+        `${documentId}.incremental.${this.#changeCount[documentId]}`,
         binary
       )
 
-      this.incrementalChanges[documentId]++
+      this.#changeCount[documentId]++
     }
   }
 
-  saveTotal(documentId: DocumentId, doc: Automerge.Doc<unknown>) {
-    const binary = Automerge.save(doc)
-    this.storageAdapter.save(`${documentId}.snapshot`, binary)
+  #saveTotal(documentId: DocumentId, doc: A.Doc<unknown>) {
+    const binary = A.save(doc)
+    this.#storageAdapter.save(`${documentId}.snapshot`, binary)
 
-    for (let i = 0; i < this.incrementalChanges[documentId]; i++) {
-      this.storageAdapter.remove(`${documentId}.incremental.${i}`)
+    for (let i = 0; i < this.#changeCount[documentId]; i++) {
+      this.#storageAdapter.remove(`${documentId}.incremental.${i}`)
     }
 
-    this.incrementalChanges[documentId] = 0
+    this.#changeCount[documentId] = 0
   }
 
-  async load(storageKey: string): Promise<Uint8Array> {
+  async loadBinary(documentId: string): Promise<Uint8Array> {
     const result = []
-    let binary = await this.storageAdapter.load(`${storageKey}.snapshot`)
+    let binary = await this.#storageAdapter.load(`${documentId}.snapshot`)
     if (binary && binary.length > 0) {
       result.push(binary)
     }
 
     let index = 0
     while (
-      (binary = await this.storageAdapter.load(
-        `${storageKey}.incremental.${index}`
+      (binary = await this.#storageAdapter.load(
+        `${documentId}.incremental.${index}`
       ))
     ) {
-      if (binary && binary.length > 0) {
-        result.push(binary)
-      }
+      if (binary && binary.length > 0) result.push(binary)
       index += 1
     }
 
     return mergeArrays(result)
   }
 
-  // TODO: make this, you know, good.
-  shouldCompact(documentId: DocumentId) {
-    return this.incrementalChanges[documentId] >= 20
+  async load<T>(
+    documentId: DocumentId,
+    prevDoc: A.Doc<T> = A.init<T>()
+  ): Promise<A.Doc<T>> {
+    return A.loadIncremental(prevDoc, await this.loadBinary(documentId))
   }
 
-  save(documentId: DocumentId, doc: Automerge.Doc<unknown>) {
-    if (this.shouldCompact(documentId)) {
-      this.saveTotal(documentId, doc)
+  save(documentId: DocumentId, doc: A.Doc<unknown>) {
+    if (this.#shouldCompact(documentId)) {
+      this.#saveTotal(documentId, doc)
     } else {
-      this.saveIncremental(documentId, doc)
+      this.#saveIncremental(documentId, doc)
     }
   }
+
+  // TODO: make this, you know, good.
+  #shouldCompact(documentId: DocumentId) {
+    return this.#changeCount[documentId] >= 20
```

Comment:
still ugh

Reply:
Using change count is probably roughly the right metric here but the number of changes between compactions is going to be determined by the per-key overhead of the underlying storage adapter. A round-trip to S3 costs a lot more than grabbing the next object from a range query in IndexedDB or just tagging on new changes to the end of a POSIX file. 

Probably a good way to improve this would be to set the change count per compaction at the StorageAdapter level... but for now, this isn't causing any pain so we can leave it be while we deal with worse problems.